### ISSUE TITLE
normal mode hardcoded with english layout

### ISSUE THREAD

**Issue by whoizit** [helix-editor/helix#133]
this bug is thousands of years old, it is present in vi / vim / nvim and in the post-modern editor it is also present. This is super sad.
It must use key codes, not key chars.

**Comment by kirawi** [comment]
To be fair, won't most programmers already be typing Latin characters?

**Comment by whoizit** [comment]
so is this a text editor or a code editor? Constantly switching the layout to get work in default mode is a pain for everyone in the world who speaks more than just English. And who writes more than just code.

**Comment by kirawi** [comment]
I believe Helix is moreso a code editor than a text editor, but you make a valid point regardless. This should definitely be a feature in the future.

**Comment by itzmjauz** [comment]
> You make a valid point, this should definitely be a feature. However, it's unclear to me why you wouldn't be able to use alternate mappings on vim/neovim/etc. I'd think this is probably the route Helix should go as well.

Alternate mappings are the sensible solution to this problem(wouldn't call it a bug), the alternative is a very bloated base configuration. Especially given the way crossterm keycode mappings work https://docs.rs/crossterm/0.19.0/crossterm/event/enum.KeyCode.html

**Comment by archseer** [comment]
As someone that uses an alternate layout too (Norman) I can sympathize with the problem. This needs some changes in crossterm: We'd need to be able to receive raw position-based key codes but crossterm only offers us key chars.

**Comment by pickfire** [comment]
I also use an alternative layout (dvorak) and use IME to type chinese and japanese. I don't see an issue here besides the invalid IME position. But yeah, maybe non-latin (but non-pictographs) are an issue. What do you suggest?

**Comment by Kethku** [comment]
I've recently been dealing with this issue on other projects. The good news (once a gui lands) is that winit will very soon have the tools to fix this problem for real (this wasn't the case a few months ago). I'm interested in helping out in this area as its fresh in my mind at least on the gui side

**Comment by pickfire** [comment]
> so is this a text editor or a code editor? Constantly switching the layout to get work in default mode is a pain for everyone in the world who speaks more than just English. And who writes more than just code.

I speak chinese and malay as well but I don't have any issue with the defaults. At least I still use a US keyboard so this must have affected those that use non-us keyboard like in europe.

I think there should be a flag. I still prefer to use key chars rather than keycode. The default keys for dvorak is still good for me to use it.

A temporary solution right now is to use the key mappings configuration. @whoizit you can map out the keys and submit it to contrib/ or something so that others is able to use it. Or maybe have something like the current theme configuration where you can just `:keymap <keys>`.

**Comment by archseer** [comment]
`showkey --scancodes` internals: an `ioctl` call can set KDSKBMODE to `RAW` for scancodes (`MEDIUMRAW` to get keycodes again)

**Comment by blaggacao** [comment]
If there is a possibility to access scancodes directly past the terminal, it would be possible to avoid having to implement [CSIu mappers](http://www.leonerd.org.uk/hacks/fixterms/) like [this one](https://github.com/nix-community/home-manager/pull/2187/files) in order get access to some key combinations.

It would be a solution at the root.

Maybe helpful:
- https://docs.rs/evdev/0.11.0/evdev/
- https://docs.rs/evdev/0.11.0/evdev/struct.Key.html

A.f.a.i.r. and as the `evdev` rust docs say:

> The Linux kernel’s “evdev” subsystem exposes input devices to userspace in a generic, consistent way.

It looks like a definite route to obviate CSIu hacks for the century-old ASCII limitations on key combinations.

**Comment by whoizit** [comment]
I am not (yet) a helix user, I have not read the thread (hard work for me, not english speaker), but how i see it
Look how it works in sway:

Bindings to keysyms are layout-dependent. This can be changed with the --to-code flag. In this case, the keysyms will be translated into the corresponding keycodes in the first configured layout.
https://github.com/swaywm/sway/wiki#key-bindings-on-a-dual-usrussian-layout
swaywm/sway#5409

so i can do even so:

set $bindsym bindsym --to-code

$bindsym {
...
    $mod+s layout stacking
    $mod+w layout tabbed
    $mod+e layout toggle split   
...
}

all I changed from the default (us-only configuration) is bindsym replaced with bindsym --to-code and began to use $bindsym variable instead bindsym keyword. And it worked no matter what layout I'm on

do not confuse bindsym and bindcode

**Comment by blaggacao** [comment]
For a text editor to become an extension of your brain through your fingers in 21st century, **we need to stop** this awful indirection going through the visual cortex (a.k.a. key char aliases) just to make keystroke decisions.

I guess a blind person would be a real good advisor, but just imagine a keyboard with blank keykaps.

In my opinion, the implementation primitives should embrace some representation of "muscle memory" locators instead. Universal libinput` scan codes are the thing that comes closest to that.

**Comment by archseer** [comment]
But you're forgetting that not all keyboards have [the same](https://olkb.com/) [physical](https://ergodox-ez.com/) [layout](https://imgur.com/a/LEhOjHg#l6epzmu). I find sway's behavior acceptable because I can bind the thumb cluster based on my default layout without having to figure out scancodes for each key.

**Comment by whoizit** [comment]
by the way, someone may not use the `us` layout at all, never

**Comment by blaggacao** [comment]
> But you're forgetting that not all keyboards have the same physical layout. 

Very true, I _really_ just wanted to complement the discussion with a radically different angle in the hopes that this angle can make `helix` a _radically_ better editor.

In terms of locators, each finger has it's own coordinate system from the resting position. So the matrix is not a parallel one but a curved one, according to the form of the hand.

Rectangular keyboards are just a very shameful violation of (human) hand ergonomics and should not be the inspiration for our primitives.

**Comment by whoizit** [comment]
just use classic keyboards by swiping diagonally across the keys and everything will be comfortable (peeped at  a4tech anti-rsi keyboards). No separate keyboards needed, no a4tech keyboard needed, just regular keyboard enough and think a little.
 
http://www.compax.ru/bib/rsi/typeme.jpg

**Comment by blaggacao** [comment]
> just use classic keyboards by swiping diagonally across the keys and everything will be comfortable

That somewhat works, but it requires elbow movements for smaller hands.

But don't get me wrong, I just want to strengthen the point why char references are bad, so that we can have layout-independent-modal commands as a collateral of a new mindset. :sweat_smile:

**Comment by whoizit** [comment]
> That somewhat works, but it requires elbow movements for smaller hands.
I've been typing for so many years, I'm fine. Prior to that, he studied academic touch typing. It was hard and painful, then I came across a rhombic keyboard and realized what I was doing wrong. This method suits me. This is applicable for any conventional keyboard. I have the keyboard on my lap, slightly diagonally. the left side of the keyboard is closer to the body, the right side is farther away. 
(translated)

**Comment by voroskoi** [comment]
Hi,

I'm just being curious: How do You handle the case with other programs? So You set up Your editor to keycodes, but every other application still uses other keys. Do You reconfigure them too? For example vim bindings are quite common, it seems quite a lot work to setup everything.

I use hungarian keyboard, searching(`/`) is SHIFT+6 for example, which is quite bad, but I haven't changed it as I have to press this for changing dirs and so too, using different keys seems strange. I could use `us` layout, but then I can not use accent keys (`éáű` etc.).

I have found https://github.com/samvel1024/kbct which seems a global solution for this problem. OTOH if I remap `/` to `é` for example, then I can not type `é` anymore, so this is just changing layout to `us`. What I would really use is something that lets me use my accent chars in *insert-mode* and use the same keys to search and so in *normal* mode, but with every program.
(Actually that still would not work, because I may have directories with accents chars in its name, so typing `~/é` would be quite difficult when `/` is bould to `é`). Am I over-complicating this?

**Comment by gagbo** [comment]
Hello,

I'm also interested in looking out how to advance on this subject, I'm using bépo (a french dvorak-like variant), and it'd be really nice if it were possible to have the possibility to use scancodes or some sort of keycode translation to describe default bindings. It would greatly reduce the amount of rebinding I have to do with editors.

For the time being, I only found the implementation in SDL, which apparently is only able to go back from keycode to scancode by [manually testing all scancode candidates](https://github.com/libsdl-org/SDL/blob/cb1e20b058f1de55b98f32b7d16cff1f896f88dd/src/events/SDL_keyboard.c#L956-L969). Because of this I think that hoping to access scancodes directly is a tough sell, and to avoid perf related issues, I guess the "way to go" would be to:

- allow keycodes to be used in the (Rust) configuration, so that default configuration is still mostly "position based" for all keyboard layouts. (Even with my layout that has `ctsr` for `hjkl`, the keycodes for `ctsr` are still `43..46`, just mapped to different keysyms)
- allow users to override bindings in TOML using either
  + "natural" notation which would describe keys just like today
  + "keycode" notation which would take the integer code values as keys for bindings
  
With this, there'd be no remapping for anyone, and non-qwerty users would have a way to get at least home-row normal mode bindings out of the box. A lot of mnemonic keys would get moved to weird places though (in my examble, **y**anking becomes `^`, **f**inding becomes `e`...), but I'd rather deal with those on a case by case basis.
  
It probably has already been proposed but I'm not sure I saw it, if I'm saying obviously bad stuff please correct me,

Regards,
Gerry

**Comment by kirawi** [comment]
I don't know if it's possible, but maybe we could provide a utility tool to do the above?

**Comment by pickfire** [comment]
A simpler alternative is rewrite the keymap for specific keybinding.

Even as a dvorak user myself I don't find this useful. Probably need to wait for someone that is interested enough in this to be fix or just create a keymap which is less effort and works well.

**Comment by blaggacao** [comment]
I do a continue to do custom build for this. thanks to `flake.nix` that's easy.

**Comment by NNBnh** [comment]
[Kitty](https://sw.kovidgoyal.net/kitty)'s [Comprehensive keyboard handling in terminals](https://sw.kovidgoyal.net/kitty/keyboard-protocol) protocol might be helpful.

**Comment by VKondakoff** [comment]
Hello! 
Are there any progress on this topic? 
Currently it is almost impossible to use Helix on a systems with two (or more) keyboard layouts: when you switch keyboard layout to a non-Latin one, the Normal mode stops to respond. 
There are several workarounds in Vim/Neovim intended to overcome this issue: langmaps, keymaps, xkb-switching. All of them have pros and cons. What are the Helix devs planning on this issue? (I hope Helix will not become an editor, which supports only Latin keyboard layout).
Thanks!

**Comment by pickfire** [comment]
Let me resend the same message.

A simpler alternative is rewrite the keymap for specific keybinding.

Even as a dvorak user myself I don't find this useful. Probably need to wait for someone that is interested enough in this to fix it or just create a keymap which is less effort and works well.

No progress on this until someone say they are working on this or someone submits a pull request. At least I don't think any of the contributors are very interested in this now, if people think it is important there should already be a patch by now.

**Comment by whoizit** [comment]
@pickfire I guess you still not understand issue, dvorak it's english layout, but this issue about non-english layout's like russian, finnish, franch, italianon and thousands overs

**Comment by VKondakoff** [comment]
> I've recently been dealing with this issue on other projects. The good news (once a gui lands) is that winit will very soon have the tools to fix this problem for real (this wasn't the case a few months ago). I'm interested in helping out in this area as its fresh in my mind at least on the gui side

Can you, please, bring a little bit more light here? What are these tools? Is something landed? Thanks!

**Comment by kirawi** [comment]
https://github.com/rust-windowing/winit/issues/753?

**Comment by pickfire** [comment]
> @pickfire I guess you still not understand issue, dvorak it's english layout, but this issue about non-english layout's like russian, finnish, franch, italiano and thousands others

Ah, now I get what you meant. Like uk or japanese keyboard (at least these are the ones I have seen) which have extra keys.

**Comment by whoizit** [comment]
no, it's not about extra keys. Youu need to switch the layout to English in order to interact with the editor, that's what this is about.

**Comment by VKondakoff** [comment]
This ⬆.

Here is a short video. I'm pressing a simple key sequence `i<Esc>v<Esc>s<Esc>` in Normal mode, while my keyboard layout is set to `En`, then I repeat the same keys when keyboard layout is set to `Ru`:

https://user-images.githubusercontent.com/230824/142721405-b179dabc-891b-4925-bfbf-bdb24790aa39.mov

**Comment by pickfire** [comment]
Ah, I get it now. I wonder if scankey if getting the keycode is really the way to go for this since my dvorak original keys worked just fine.

How did other editors solve this, like vim or kak? Did they add in the default keys for other non-latin keymap as well?

**Comment by VKondakoff** [comment]
As far as I can say there are three main methods of fixing this issue:

1. [Langmaps](https://neovim.io/doc/user/russian.html): set langmaps to be able to enter Normal commands without switching the keyboard layouts.
`:set langmap=ФИСВУАПРШОЛДЬТЩЗЙКЫЕГМЦЧНЯ;ABCDEFGHIJKLMNOPQRSTUVWXYZ,
    фисвуапршолдьтщзйкыегмцчня;abcdefghijklmnopqrstuvwxyz`
2. [Keymaps](https://neovim.io/doc/user/russian.html): switch between keyboards independently from a system-wide keyboard settings.
`:set keymap=russian-jcukenmac`
3. [Auto-switch the system keyboard to English layout](https://github.com/lyokha/vim-xkbswitch) when leaving Insert mode and switch back to the layout which was used when you enter Insert mode again.

**Comment by NNBnh** [comment]
> 3. [Auto-switch the system keyboard to English layout](https://github.com/lyokha/vim-xkbswitch) when leaving Insert mode and switch back to the layout which was used when you enter Insert mode again.

Out of these three, this is the only option that work with Chinese, Japanese and Vietnamese.

**Comment by pickfire** [comment]
> Out of these three, this is the only option that work with Chinese, Japanese and Vietnamese.

But it needs integration with IME like https://github.com/vim-scripts/fcitx.vim and it's not usually the default for CJK. Also, non CJK users probably don't use an IME like in the case of russian which makes it not so useful for non-CJK cases.

**Comment by eugenesvk** [comment]
> > 5\. [Auto-switch the system keyboard to English layout]

> Out of these three, this is the only option that work with Chinese, Japanese and Vietnamese.

but unfortunately it introduces some mess in all the others, if I'm using several apps and switch to a non-English layout to copy/edit/paste some text in Helix and a browser, I wouldn't want Helix to mess up my system layout on context switches, because then I'd lose the predictable state in my browser (also the flashes of the language indicator aren't great)

**Comment by pickfire** [comment]
> but unfortunately it introduces some mess in all the others, if I'm using several apps and switch to a non-English layout to copy/edit/paste some text in Helix and a browser, I wouldn't want Helix to mess up my system layout on context switches, because then I'd lose the predictable state in my browser (also the flashes of the language indicator aren't great)

IME usually have the option to set state per application, so if you use a language in terminal, you can use different language in browser.

**Comment by eugenesvk** [comment]
> IME usually have the option to set state per application

Ok, so the inelegant mess is only introduced to those users that don't have this setting enabled :) (I don't as I found it more confusing than helpful)

Also, I have many other tabs in the terminal, so if I'm typing in a non-English Helix Insert mode and switch to another tab, the language would presumably not switch back to English as I haven't quit the Insert mode yet, right?

**Comment by pickfire** [comment]
> Also, I have many other tabs in the terminal, so if I'm typing in a non-English Helix Insert mode and switch to another tab, the language would presumably not switch back to English as I haven't quit the Insert mode yet, right?

Yes, it won't but I never had the need to do that.

**Comment by whoizit** [comment]
https://github.com/aveplen/ruscmd.nvim

**Comment by nanoqsh** [comment]
Helix is a great editor and I'm sure it's the future. But I'm saddened by the non-support of different keyboard layouts out of the box. Although vim has the same problem, I expect a more modern editor to fix this. Similar to how Rust supports utf-8 compared to C and I no longer have to think about special handling of non-latin texts.

But how to do that? As I understand it, supporting different layouts requires different solutions. You can give the ability to set keys by specifying keycodes, but how convenient will it be for the case of different language layouts?
```toml
[keys.normal]
"34" = { "1c" = "code_action" }
```

Users just want to specify mappings like `C-s` and have it just work. At the same time, without duplicating it like this:
```toml
[keys.normal]
"C-s" = ":w"
"C-ы" = ":w"
```

So why not do something like vim's "langmaps"? When I press a key (not in insert mode) the editor will try to resolve the input to the given "lang" mappings:
```toml
[langmapping]
"ы" = "s"
"Ы" = "S" # Perhaps this duplication can be avoided
"ф" = "a"
# ..
```

So (almost) all previously set mappings will work. For most users this feature will be enough until key codes supporting is implemented. For many languages a default configuration can be provided, which will only need to be enabled: `langmapping = "cyrillic"`. Make a directory from where the corresponding `.toml` file will be imported, how it works with themes now.

Unfortunately, this solution is not absolute. Different layouts don't match not only in letters, but also in symbols too. For example, the Shift + 2 enters either `@` or `"` depending on the layout. So this problem is much more fundamental and rests on the initially crookedly designed layout.

That's pretty much how I see it. Perhaps this function should come up with a more descriptive name.

**Comment by whoizit** [comment]
All existing solutions offered in (n)vim editor are crutches, they are either incomplete and still make you uncomfortable, or do not work in `wayland`. 

Also, entering command mode should **always** switch the layout to English, because all commands in command mode are entered in English, **always**

In the Windows operating system, this problem did not exist for me at all. All hotkeys always worked the same in both languages in any program. But I haven't used Windows for 15 years and it's a matter of principle.

But Linux in the 21st century still has great programs that only work properly for english-native speakers (and it looks like for CJK-guys). This programs like (n)vim and mplayer/mpv

We simply have to get rid of the steamy inheritance and rewrite these programs

**Comment by pickfire** [comment]
@whoizit I believe you can try solving this first by using adding a contrib file to map the keys.

I wonder if it is good to have a set of translation, since it may not always be a one-on-one translation for the keys, if a language have lesser keys than latin then it means that some keys could not be pressed, so I think it might be good to have the keys being remapped completely.

**Comment by whoizit** [comment]
good solution from Sway WM https://wiki.archlinux.org/title/Sway#Keybindings_and_keyboard_layouts

**Comment by s0LA1337** [comment]
Another approach could be to provide some kind of matrix, which then internally will remap the keys based on the location within the matrix.
Unfortunately i am not sure how to express this in toml but in general i mean something like this:
```toml
[editor.keys.matrix]
matrix = [
"qwfpgjluy;[]",
"arstdhneio'",
"zxcvbkm,./"
]
```

something like this or alike.
this could resolve something for other people (like me, using colemak) in a rather simple way.

**Comment by VKondakoff** [comment]
At least right now trying to add Cyrillic letter to 'config.toml' fires an error 
```Bad config: unexpected character found: `\u{444}` at line 2 column 1
Press <ENTER> to continue with default config```

Here is a test config:
```
[keys.normal]
ф = "move_char_left" # Maps the 'a' key to the move_char_left command
ц = "move_line_up" # Maps the 'w' key move_line_up
[keys.insert]
j = { k = "normal_mode" }
```

**Comment by VKondakoff** [comment]
Thanks, it works with the quotes like this:
```
[keys.normal]
"р" = "move_char_left" # Maps the 'р' key to the move_char_left command
"л" = "move_line_up" # Maps the 'л' key to move_line_up command
"д" = "move_char_right" # Maps the 'д' key to the move_char_right command
"о" = "move_line_down" # Maps the 'о' key to move_line_down command
[keys.insert] 
"о" = { "л" = "normal_mode" }
"j" = { "k" = "normal_mode" }
```

BTW, there is nothing about  quotes in the [docs](https://docs.helix-editor.com/remapping.html):
```
# At most one section each of 'keys.normal', 'keys.insert' and 'keys.select'
[keys.normal]
C-s = ":w" # Maps the Control-s to the typable command :w which is an alias for :write (save file)
C-o = ":open ~/.config/helix/config.toml" # Maps the Control-o to opening of the helix config file
a = "move_char_left" # Maps the 'a' key to the move_char_left command
w = "move_line_up" # Maps the 'w' key move_line_up
"C-S-esc" = "extend_line" # Maps Control-Shift-Escape to extend_line
g = { a = "code_action" } # Maps `ga` to show possible code actions
"ret" = ["open_below", "normal_mode"] # Maps the enter key to open_below then re-enter normal mode

[keys.insert]
"A-x" = "normal_mode" # Maps Alt-X to enter normal mode
j = { k = "normal_mode" } # Maps `jk` to exit insert mode

```

**Comment by edio** [comment]
> BTW, there is nothing about quotes in the [docs](https://docs.helix-editor.com/remapping.html):

you are looking at the wrong docs https://toml.io/en/v1.0.0#keys

> Bare keys may only contain ASCII letters, ASCII digits, underscores, and dashes (A-Za-z0-9_-).

> Quoted keys follow the exact same rules as either basic strings or literal strings and allow you to use a much broader set of key names.

**Comment by jgirardet** [comment]
Hi,
French layout here.
I wish suggest a little hint  a small trick could to remapa key  to another one in normal mode.
For example:
the unimpaired works all with [ and ] which aren't accessible at all for us.
If I could remap "ù" to "[" in normal mode, I could then use all unimpared for free withtout having to configure all shortcuts by hand.

**Comment by Shirk** [comment]
Just another voice for a *properly implemented* langmap-like feature.
Contrary to the comments about it being a crutch.. it's not if you use it correctly and it works on any WM and any platform that has a vim.

`langmap` is accompanied by an additional command called `langnoremap` in vim/nvim. The later being responsible for preventing keys issued by plugins to be remapped. With that you can very easily fix your non-insert-mode to work for most layouts without breaking functionality.

```vim
" swap j/k/l keys with d/r/n to get movement keys on German Dvorak
set langmap=dj,rk,nl,DJ,RK,NL,jd,kr,ln,JD,KR,LN
" don't override mappings for plugin generated input
set langnoremap
```

**Comment by vwkd** [comment]
+1

VSCode gets this right.

Currently, helix needs an additional two keystrokes for every input to switch to / from foreign language.

**Comment by idoric** [comment]
I have read this discussion carefully, and I would like to add my personal testimony. Like @gagbo I'm using the bépo layout (some kind of french dvorak), and I want to give it a try to helix. After thinking about it, I came to the conclusion that the easiest way for me was to use the qwerty us layout in normal mode, and the bépo layout in insert mode¹. Since there's no equivalent of Vim's langmap, I wanted to remap all the commands (from documentation keymap page) to simulate a qwerty us keyboard under the [keys.normal] section of config.toml. Unfortunately I soon found that this wasn't possible to do it in a consistent way, starting with the basic command "repeat last insert", which cannot be remapped.

(Moreover, in the case of the dot to repeat the last insertion, it is also very annoying in azerty, the most widespread keyboard in France, because the dot is not directly accessible, you have to press shift in same time.)

¹ For the more curious, I let you see where the h, j, k and l keys are on a bépo keyboard: https://bepo.fr/ (the layout is displayed at the top of the home page). And that's far from the only difficulty, you can also search for [ and ] as a second example.

**Comment by Shirk** [comment]
> to be honest i would prefer for the command mode to use the scancode or keyid instead of remappings. swiss german or swiss french keyboards have a dead key on the english ] position.
> 
> something in the lines of this tiny library. https://crates.io/crates/keyboard_query

Scancodes don't work well at all for anything that is not a level-zero (direct input) key as you don't have an easy way to infere what scancode + modifier (shift, alt, ctrl, super, meta, option) maps to because that's something the OS keyboard layout is responsible for. And this doesn't even account for modifier combinations or the famous ISO-Level3 shift (alt-gr) that German and so many of our European languages have and which are the only way to access things like {[]}@.

**Comment by soloturn** [comment]
> Scancodes don't work well at all for anything that is not a level-zero (direct input) key as you don't have an easy way to infere what scancode + modifier (shift, alt, ctrl, super, meta, option) maps to because that's something the OS keyboard layout is 

the beautiful thing with helix, vi is that it does NOT use modified keys on the command mode. for entering text a different mode is used fortunately, i am happy how it behaves now. 

> _EDIT_ in addition to this it also completely ignores the idea of being able to remap _some keys_ but not all - going by scancodes entering _I_nsert mode on a dvorak layout would - for example - require pressing 'c' (which is the physical 'i' key) and that makes mnemonics completely useless.
true, mnemonics is gone for a couple of characters, for others like the "hjkl" key navigation it was never there. its programmed like a terminal game - easy to type - and this is good so :) remapping scancodes should be as well possible, it is just a parameter?

**Comment by Shirk** [comment]
> > Scancodes don't work well at all for anything that is not a level-zero (direct input) key as you don't have an easy way to infere what scancode + modifier (shift, alt, ctrl, super, meta, option) maps to because that's something the OS keyboard layout is
> 
> the beautiful thing with helix, vi is that it does NOT use modified keys on the command mode. for entering text a different mode is used fortunately, i am happy how it behaves now. remapping scancodes should be as well possible, it is just a parameter?

Well using keycodes for input mode but scancodes for everything else sounds like a great mess and rather complicated to implement requiring two models of requesting input from the OS for each platform helix runs on..

> vi / helix command mode reminds me on a terminal game - easy to type - and this is good so :) i d love to see the person first, who uses dvorak and insists on mnemonics. true, the i - insert mnemonics is gone on a dvorak, for others like the "hjkl" key navigation mnemonics was always the terminal game mnemonics. that one is masochistic to type on dvorak btw.

Well you found one.. and honestly, mnemonics are key for all modal editors like vi, vim, nvim, helix and kakoune unless you are one of the persons that constantly looks at the letters written on the keys while typing. Call me wrong but 'c' doesn't really associate mentally with 'Insert', does it? Or 'uh' for 'go home / start of line', neither would ',' make sense to move words or 'e' for "delete" or 'q' to select the current line - all of these rely on mnemonics and don't work with scancode-based command modes unless you remap every key at which point you don't need scancodes anymore. 

(P.S. all of these examples are the scancodes that correspond to 'i', 'gh', 'w', 'd' and 'x' respectively when typing on a dvorak-DE layout. All of these work as expected with something like langmap / langnoremap in vim and nvim without needing scancodes)

**Comment by ZabejAga** [comment]
There is a solution for the (n)vim/kak/VScodeVIM editors : plugins like `im-select`.
It will change layout to defined (i.e. 'US') when you change mode to normal. 
Would be nice to have such solution here as it solves(at least part of) the issue.

**Comment by Shirk** [comment]
> There is a solution for the (n)vim/kak/VScodeVIM editors : plugins like `im-select`. It will change layout to defined (i.e. 'US') when you change mode to normal. Would be nice to have such solution here as it solves(at least part of) the issue.

Sadly also only a workaround at best - as a foreign keyboard layout user you'd probably still only want to remap certain keys to the "intended places" (think where hjkl is) but keep the others where your layout has them so you can blind type the mnemonics without having to switch your brain back and forth between two layouts.

One workaround I've found that works for my DVORAK-DE layout - but it's wanting as it takes away some of the Alt+Key combinations - is to use Alt + H/D/R/N (the layout keys I get when pressing the physical h/j/k/l) and remap them to the cursor move and selection commands like this, but it's more of a hack than a real fix :|

```toml
[keys.normal]
"A-h" = "move_char_left"
"A-d" = "move_line_down"
"A-r" = "move_line_up"
"A-n" = "move_char_right"

[keys.select]
"A-h" = "extend_char_left"
"A-d" = "extend_visual_line_down"
"A-r" = "extend_visual_line_up"
"A-n" = "extend_char_right"
```

That way does provide movement my holding the left alt key and pressing the physical h/j/k/l but it also uses up those alt+key combos.  The one advantage over a complete layout change this has is that I can still use 'd' to 'd'elete while alt+d (which is the physical 'j' key) moves my cursor etc.

**Comment by ZabejAga** [comment]
> Sadly also only a workaround at best - as a foreign keyboard layout user you'd probably still only want to remap certain keys to the "intended places" (think where hjkl is) but keep the others where your layout has them so you can blind type the mnemonics without having to switch your brain back and forth between two layouts.

im-select is helpful when you mostly use non-latin layouts (I'm not programmer but notetaker).

 
> One workaround I've found that works for my DVORAK-DE layout - but it's wanting as it takes away some of the Alt+Key combinations - is to use Alt + H/D/R/N (the layout keys I get when pressing the physical h/j/k/l) and remap them to the cursor move and selection commands like this, but it's more of a hack than a real fix :|

for such case I use remapped CAPSLOCk to Hyper and 'Hyper + <homerow_keys>' as arrows.

**Comment by eugenesvk** [comment]
By the way, before Helix does the proper thing and implements a more convenient native support for alternative keyboard layouts you could use a couple of Python scripts to extend/replace your config to an alternative keymap

https://github.com/helix-editor/helix/discussions/7097

**Comment by eugenesvk** [comment]
> There is a solution for the (n)vim/kak/VScodeVIM editors : plugins like `im-select`

But that's not a solution, e.g., if you have a visible keyboard selection indicator, it will constantly flash despite the fact that you don't really change any layout

**Comment by eugenesvk** [comment]
> Call me wrong but 'c' doesn't really associate mentally with 'Insert', does it?

neither does `h` associate with `left`. Mnemonics aren't useful for such frequent keys as cursor movements or switching to insert mode, your muscle memory will very quickly map it to a finger, not a keycap (thus it's best to use a better key position), so being able to have consistent physical mappings is a plus of scancodes

But you're right that it should be flexible so that people who value mnemonics over physical location could continue to do so

By the way, WezTerm has this cool feature that allows you to pick the mode on a per-key basis

> You can explicitly assign using the physical position by adding a phys: prefix to the value, for example: key="phys:A". This will match key presses for the key that would be in the position of the A key on an ANSI US keyboard.

https://wezfurlong.org/wezterm/config/keys.html#physical-vs-mapped-key-assignments

**Comment by Shirk** [comment]
Well the point of <key>h</key>/<key>j</key>/<key>k</key>/<key>l</key> is indeed to become "muscle memory" but it also mirrors the layout of arrow keys on older terminal keyboards and is mainly intended to keep your right hand on the homerow instead of having to move it all the time. 

*However* literally all other keys are mnemonics matching the action they perform.
The whole reason being that it helps you learn the action by deducing the keys from what you want to do.

Not using keys in your native layout is extremely crippling to your muscle memory training because it forms from associating things like "**I**nsert" or "**G**o **H**ome" or "**D**elete" - an action with a matching letter - with a muscle movement. Things which you sabotage by having zero relation between the key, and the action it triggers.
 
Alas.. we're derailing the topic.

**Comment by eugenesvk** [comment]
> mainly intended to keep your right hand on the homerow instead 

Same principle - you chose a more convenient location instead of relying on a keycap match

> Not using keys in your native layout is extremely crippling to your muscle memory training because it forms from associating things like "**I**nsert" or "**G**o **H**ome" or "**D**elete" - an action with a matching letter - with a muscle movement.

you're not crippling it for the same reason `jkl` does not - keycap associations don't provide much over repetition, which is plenty for these commonly used keys like Insert. And then there are too  many commands that start with some letter for you to have a unique association (how does `D` delete, left or right? with or without `yank`, where would `decrement` command go?)

>  Things which you sabotage by having zero relation between the key, and the action it triggers.

The relation is not zero, it's just not keycap-based. For example, `H` is positioned to the left, and home is positioned to the left, so here is your association - you move your finger off home keys to the left to move further left (or I'm using `h` as backspace with a similar direction-based mnemonic)

> Alas.. we're derailing the topic.

Well, scan codes are a solution to this very topic, and they are a better solution than the workaround I've posted above since they also work for symbol keys and don't pollute the popup window, but you're arguing against them based on the value of keycap-based mnemonics

**Comment by alexeymuranov** [comment]
An idea: #9896.

**Comment by HealsCodes** [comment]
> An idea: #9896.

While this is a valid idea it would still require that 'minor' normal mode to use scancode based input or translate the keys to scancodes on the fly in order to be able to act on key positions instead of the received character.

Otherwise youre asking to provide a mapping for every possible keyboard layout in existence and any that might crop up in the future or are in use by people that make their own.

That's pretty much exactly why ViM has langmap which was suggested a number of times already.. it allows you to switch an arbitrary number of pairs of keys but only while normal mode is active.

**Comment by alexeymuranov** [comment]
@HealsCodes, I do not mind some kind of configured automated `langmap` for my both 'major' and 'minor' normal modes. As long as it works. (Which is not the case in Vim: [vim/issues/3018], [vim/issues/5147].)

[vim/issues/3018]: https://github.com/vim/vim/issues/3018
[vim/issues/5147]: https://github.com/vim/vim/issues/5147

**Comment by HealsCodes** [comment]
"not the case" is a wide stretch - I've been using it in ViM for over a decade and it works in combination with langnoremap.
Not surprised it wasn't implemented with multibyte characters back in the day (you have to account for the age of ViM, proper unicode support came way after). 

If you were to implement this feature in recent days you would probably use unicode-codepoints or actual scancodes for the langmap syntax to begin with ^^'
